neondatabase/neon
# pageserver: add vectored get implementation

## Summary
This PR introduces a new vectored implementation of the read path.

The general algorithm is describe in code comments, but I'll add it here too for visibility:

The search maintains a fringe (`LayerFringe`) which tracks all the layers that intersect
the current keyspace being searched. Continuously pop the top of the fringe (layer with highest LSN)
and get all the data required from the layer in one go.

More granularly, the algorithm is as follows:
1. Check if all keys are done and stop the search if true
2. Split the unmapped key space up to account for gaps left by finished keys
3. Descend into the next timeline if needed
4. Find the next layers for the unmapped keyspace. This can be open, frozen or disk and uses the layer map ranged search.
5. Pop the layer at the top of the fringe and grab all the required data from it
6. Mark the ranges that previously mapped to this layer as unmapped
7. Go back to 1

Apart from the high level layer traversal, vectored variants have been introduced for grabbing
data from each layer type. They still suffer from read amplification issues and that will be addressed
in a different PR.

You might notice that in some places we duplicate the code for the existing read path. All of that code
will be removed when we switch the non-vectored read path to proxy into the vectored read path.
In the meantime, we'll have to contend with the extra cruft for the sake of testing and gentle releasing.

## Reviewing
I would review this by commit. Put quite a bit of effort into keeping them atomic and small.

## Testing & Release Plan
We'll need to build some confidence in these changes before using them on real pageservers.
This PR adds a config flag, `get_vectored_impl`, which defaults to the sequential implementation
that we are using in the fleet now.
 
My plan is as follows, but it's not set in stone so please feel free to suggest things. 
1. Set `get_vectored_impl` to `vectored` to exercise the new code ***only*** in CI regress tests. In debug builds, we cross validate the sequential and vectored searches.
2. Update `pagebench` to issue vectored gets. This should build more confidence in the change and give us some idea
of how perf looks like.
3. Enable the config flag in staging gradually (potentially only for SLRU fetches and then for compaction to minimise risk of data corruption).
4. Enable the config flag in prod gradually.

## Checklist before requesting a review

- [ ] Do we need to implement analytics? if so did you add the relevant metrics to the dashboard?
- [ ] If this PR requires public announcement, mark it with /release-notes label and add several sentences in this section.

## Checklist before merging

- [ ] Do not forget to reformat commit message to not include the above checklist
